### ISSUE TITLE
Virtual environment created by Python 3 venv module not detected by use_virtualenv

### ISSUE THREAD

**Issue by chmue** [rstudio/reticulate#194]
I have a python3 virtualenv residing in `~/src/venv` which I created with (on Linux):
```sh
cd ~/src
python3 -m venv venv
```
I cannot set reticulate to this virtual environment with `use_virtualenv()` at the moment:
```
> library("reticulate")
> use_virtualenv("~/src/venv", required = TRUE)
Error in use_virtualenv("~/src/venv", required = TRUE) : 
  Directory ~/src/venv is not a Python virtualenv
```
The problem seems to be that `use_virtualenv` is looking for a file called `activate_this.py` in `VIRTUALENV/bin` which does not exist ([in this line](https://github.com/rstudio/reticulate/blob/3887044186853bf2666b6ba32a55f2ddbea7be61/R/use_python.R#L50)):
```
$ cd ~/src/venv/bin
$ ls -1
activate
activate.csh
activate.fish
chardetect
dill_unpickle.py
easy_install
easy_install-3.6
f2py
ftfy
get_objgraph.py
pip
pip3
pip3.6
plac_runner.py
__pycache__
pyperf
python
python3
spacy
tqdm
```

Does this mean that virtual environments created by the [Python 3 venv module](https://docs.python.org/3/library/venv.html) are not supported or am I missing something?

**Comment by jjallaire** [comment]
That's correct, they are not supported. The reason we need `activate_this.py` as opposed to an activate script is that we are an already running process as opposed to a shell environment, so we need the "activation" (which sets various environment variables and manipulates the PATH) to be done in Python code. There's nothing that says we couldn't find a way to make this work (don't know enough about the venv module to say) but it definitely won't work right now.

**Comment by pagepiccinini** [comment]
Any timelines on when python3 virtualenv will be supported, or suggested work arounds until then?

**Comment by jjallaire** [comment]
We need a way to initialize the virtualenv from Python (as opposed to from the shell). It looks like from here that there is no equivalent of activate_this.py for Python 3 virtualenvs: https://stackoverflow.com/questions/25020451/no-activate-this-py-file-in-venv-pyvenv

So I guess we need a way to emulate activate_this.py. Don't have time to dig into this right now but would certainly welcome a PR.

**Comment by chmue** [comment]
I looked into this for my initial question:

The [documentation](https://docs.python.org/3/library/venv.html#venv-def) for the `venv` module states that the two variables, `sys.prefix` and `sys.exec_prefix`, need to be set.

From reading [PEP 405](https://www.python.org/dev/peps/pep-0405/) I understand that if the Python executable symlinked to the virtual environment is started, Python will do the bootstrapping of the virtual enviroment (setting `sys.prefix` and `sys.exec_prefix`) on its own. So maybe there is no special logic needed in reticulate as long as the symlinked Python binary in the virtual environment folder is started.

Maybe it would be possible to check for a `pyvenv.cfg` file in the main directory of the virtual environment to detect whether the venv module is used over other Python virtual environment solution.

I will run some more tests to see if this might work.

**Comment by jjallaire** [comment]
We don't actually start the Python binary (as that would create another process, eliminating the possibility of the tight interlanguage integration we have). Rather, we call the Python embedding API so are effectively stepping into the shoes of the Python binary. So we need to emulate exactly what it is doing.

**Comment by chmue** [comment]
I see, thank you very much for this information! I am unfortunately not very well-versed in C++ let alone the Python embedding API. I will still keep looking into it and keep you apprised.

**Comment by chmue** [comment]
I added a commit in my fork of reticulate which addresses this problem: https://github.com/chmue/reticulate/commit/2f2654ed966096aa674a7cc62c16721fafefb751

PEP 405 introduced "automatic" detection and setup of Python virtual environments. Therefore, `activate_this.py` does not need to be called in order to "activate" the virtual environment. Activation is performed by the Python executable directly, if it detects a PEP 405-style virtual environment. This works equally when embedding Python and calling Python directly.

This very simple solution, which just allows (symlinked) Python executables in PEP 405-style virtual environments, "just works" for my setup. I have not done any testing beyond that but would be happy to write tests or a smarter check if that would make sense for you.

Availability of the `venv` module in the Python standard library coincided with support of the Python executable for the "automatic" loading of the virtual environment. So, strictly speaking, you should not be able to create virtual environments like this in Python versions where the automatic loading is not supported. However, I am sure that there will be some edge cases which you might want to catch in an implementation.

tl;dr Support for PEP 405/Python 3.3+ virtual environments is possible and does not require changes in your Python C-API integration whatsoever. However, you might want to do smarter checking in `use_virtualenv()` than what I propose above to detect and handle edge cases and ensure that the virtual environment is loaded.

**Comment by Peder2911** [comment]
I am not getting any kind of warning when attempting this; i took me a pretty long time to figure out that it was _missing support_ that was the issue. Perhaps also a good idea to add until this is fixed?

**Comment by jjallaire** [comment]
We just merged a fix for this issue: https://github.com/rstudio/reticulate/pull/370
